File: linked-list/python/linkedlisttest.py
from linkedlist import LinkedList
from testfixtures import ShouldRaise

ll = LinkedList()

# Created empty list

assert(ll.head != None) # There is one empty node
assert(ll.head.data == None)
assert(ll.size() == 0)
assert(ll.empty())

# Insert some data bois

ll.push_back(0)
ll.push_back(1)
ll.push_back(2)
ll.push_back(3)
ll.push_back(4)

assert(ll.size() == 5)
assert(not ll.empty())
assert(ll.value_at(0) == 0)
assert(ll.value_at(1) == 1)
assert(ll.value_at(2) == 2)
assert(ll.value_at(3) == 3)
assert(ll.value_at(4) == 4)
with ShouldRaise(IndexError):
  ll.value_at(5)
ll.push_back(752)
ll.push_back(754)
ll.push_back(1020)
ll.push_back(7879)
for i in range(ll.size()):
  ll.value_at(i)
assert(ll.value_at(8) == 7879)
assert(ll.value_at(7) == 1020)
assert(ll.value_at(6) == 754)
assert(ll.value_at(5) == 752)
with ShouldRaise(IndexError):
  ll.value_at(9)

ll.push_front(-7)
assert(ll.size() == 10)
assert(ll.value_at(0) != 0)
assert(ll.value_at(0) == -7)
assert(ll.value_at(1) == 0)

with ShouldRaise(IndexError):
  ll.value_at(10)

ll.pop_front()
assert(ll.size() == 9)
assert(ll.value_at(0) != -7)
assert(ll.value_at(0) == 0)
assert(ll.value_at(1) == 1)
with ShouldRaise(IndexError):
  ll.value_at(10)  
with ShouldRaise(IndexError):
  ll.value_at(9)

ll.pop_front()
assert(ll.size() == 8)
assert(ll.value_at(0) != -7)
assert(ll.value_at(0) != 0)
assert(ll.value_at(0) == 1)
assert(ll.value_at(1) == 2)
with ShouldRaise(IndexError):
  ll.value_at(10)  
with ShouldRaise(IndexError):
  ll.value_at(9)
with ShouldRaise(IndexError):
  ll.value_at(8)

tmp_ll = LinkedList()
assert(tmp_ll.pop_front() == None)
del tmp_ll

assert(ll.front() == 1)
assert(ll.back() == 7879)

assert(ll.pop_back() == 7879)
assert(ll.size() == 7)

File: linked-list/scala/src/test/scala/LinkedListTest.scala
import main.scala.LinkedList
import org.scalatest._

class LinkedListTest extends FlatSpec with Matchers with OptionValues with Inside with Inspectors{
  "A linked list" should "add values to the end" in {
    val ll = new LinkedList[Int]
    var i = 0
    val rnd = new scala.util.Random
    while(i < 10000) {
      val random = rnd.nextInt()
      ll.push_back(random)
      i += 1
      assert(ll.back() == random)
    }
    assert(!ll.empty)
    ll.push_back(12)
    assert(ll.back() == 12)
    ll.push_back(75)
    ll.push_back(90)
    assert(ll.back() == 90)
  }

  "A linked list" should "be empty when created" in {
    val ll = new LinkedList[Int]
    assert(ll.empty)
  }

  "A linked list" should "not be empty after added elements" in {
    val ll = new LinkedList[Int]
    assert(ll.empty)
    ll.push_back(-120)
    assert(!ll.empty)
  }

  "A linked list" should "have a data in the head" in {
    val ll = new LinkedList[Int]
    assert(ll.front() == 0)
    ll.push_back(-12)
    assert(ll.front() == -12)
    assert(ll.back() == -12)
    var i = 0
    val rnd = new scala.util.Random
    while(i < 10000) {
      val random = rnd.nextInt()
      ll.push_back(random)
      i += 1
    }
    assert(ll.front() == -12)
    assert(ll.back() != -12)
  }

  "A linked list" should "have size" in {
    val ll = new LinkedList[Int]
    assert(ll.size == 0)
    ll.push_back(-12)
    assert(ll.size == 1)

    val ll2 = new LinkedList[Int]
    var i = 0
    val rnd = new scala.util.Random
    while(i < 100000){
      ll2.push_back(rnd.nextInt())
      i += 1
    }
    assert(ll2.size == 100000)
  }

  "A linked list" should "get values from different indexes" in {
    val ll = new LinkedList[Int]
    val thrown = intercept[IndexOutOfBoundsException] {
      assert(ll.value_at(0) == 0)
    }

    ll.push_back(-12)
    assert(ll.value_at(0) == -12)
    var i = 0
    val rnd = new scala.util.Random
    while(i < 100000){
      val random = rnd.nextInt()
      ll.push_back(random)
      i += 1
      assert(ll.value_at(i) == random)
      assert(ll.value_at(0) == -12)
    }
  }

  "A linked list" should "remove from the first node" in {
    val ll = new LinkedList[Int]
    print(ll.pop_front())
  }
}
